clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -0,0 +1,128 @@
+use clippy_utils::diagnostics::span_lint;
+use if_chain::if_chain;
+use rustc_ast::ast::{BlockCheckMode, ExprKind, UnsafeSource};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{source_map::SourceMap, BytePos, FileName, Span};
+use std::convert::TryInto;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `unsafe` blocks without a `// SAFETY` comment
+    /// explaining why the unsafe operations performed inside
+    /// the block are safe.
+    ///
+    /// ### Why is this bad?
+    /// Undocumented unsafe blocks are hard to check and maintain.
+    /// On the other hand, writing safety comments helps uncovering
+    /// unsoundness holes and bugs.
+    ///
+    /// ### Example
+    /// ```rust
+    /// # use std::ptr::NonNull;
+    /// # let a = &mut 42;
+    ///
+    /// let ptr = unsafe { NonNull::new_unchecked(a) };
+    /// ```
+    /// You should explain why calling `NonNull::new_unchecked` is safe:
+    /// ```rust
+    /// # use std::ptr::NonNull;
+    /// # let a = &mut 42;
+    ///
+    /// // SAFETY: references are guaranteed to be non-null.
+    /// let ptr = unsafe { NonNull::new_unchecked(a) };
+    /// ```
+    pub UNDOCUMENTED_UNSAFE_BLOCKS,
+    restriction,
+    "unsafe blocks without safety comments"
+}
+
+struct SafetyComment {
+    hi: BytePos,
+    hi_line: usize,
+}
+
+#[derive(Default)]
+pub struct UndocumentedUnsafeBlocks {
+    safety_comments: FxHashMap<FileName, Option<Vec<SafetyComment>>>,
+}
+
+impl UndocumentedUnsafeBlocks {
+    fn safety_comments(&mut self, sm: &SourceMap, span: Span) -> Option<&mut Vec<SafetyComment>> {
+        let file = sm.span_to_filename(span);
+        self.safety_comments
+            .entry(file.clone())
+            .or_insert_with(|| Self::gather_safety_comments(sm, &file))
+            .as_mut()
+    }
+
+    // Inspired by `rustc_ast::utils::comments::gather_comments`.
+    fn gather_safety_comments(sm: &SourceMap, file: &FileName) -> Option<Vec<SafetyComment>> {
+        let source_file = sm.get_source_file(file)?;
+        let src = source_file.src.as_deref()?;
+        let start_bpos = source_file.start_pos;
+        let mut comments = Vec::new();
+
+        let mut inside_comment = false;
+        let mut pos = rustc_lexer::strip_shebang(src).unwrap_or(0);
+        for token in rustc_lexer::tokenize(&src[pos..]) {
```

Comment:
Since we don't really care about correct and complete syntax in the file, but only finding comments immediately above the `unsafe` block, you should be able to optimize this. I think you can use the [`get_enclosing_scope`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.get_enclosing_scope) in combination with the [`opt_span`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.opt_span) methods and then slice the to-tokenize `src` with the start of the enclosing scope up until the start of the block.

To do this, you would have to turn this lint into a `LatePassLint`, but that is a tradeoff I'm willing to take in order to not re-lex whole files just to find a comment, of which we already know where it _should_ be.

To make sure that just slicing the `src` works as expected, please add a test case with unicode chars in the comment and the code above the `unsafe` block. 

---

With this change, you also might be able to just directly return if the checked block has a `SAFETY` comment right above it, without having to do `BytePos` and `span.hi()/.lo()` gymnastics. Basically you have to track if after lexing the suggested part of the file, the last thing you saw was a comment and if it was a `SAFETY` comment.